pkg/search/postgres/store.go

```diff
@@ -468,4 +580,226 @@ func GloballyScopedUpsertChecker[T any, PT unmarshaler[T]](targetResource permis
+		}
+		return nil, missed, nil
+	}
+	s.cacheLock.RLock()
+	defer s.cacheLock.RUnlock()
+	results := make([]PT, 0, len(identifiers))
+	misses := make([]int, 0, len(identifiers))
+	for ix, id := range identifiers {
+		val, found := s.cache[id]
+		found = found && s.isReadAllowed(ctx, val)
+		if found {
+			results = append(results, val)
+		} else {
+			misses = append(misses, ix)
+		}
+	}
+	return results, misses, nil
+}
+
+func (s *GenericStore[T, PT]) getIDsFromCache(ctx context.Context) ([]string, error) {
+	if !s.useCache {
+		return nil, nil
+	}
+	s.cacheLock.RLock()
+	defer s.cacheLock.RUnlock()
+	results := make([]string, 0, len(s.cache))
+	for k, v := range s.cache {
+		if !s.isReadAllowed(ctx, v) {
+			continue
+		}
+		results = append(results, k)
+	}
+	return results, nil
+}
+
+func (s *GenericStore[T, PT]) getAllFromCache(ctx context.Context) ([]PT, error) {
+	if !s.useCache {
+		return nil, nil
+	}
+	s.cacheLock.RLock()
+	defer s.cacheLock.RUnlock()
+	results := make([]PT, 0, len(s.cache))
+	for _, v := range s.cache {
+		if !s.isReadAllowed(ctx, v) {
+			continue
+		}
+		results = append(results, v)
+	}
+	return results, nil
+}
+
+func (s *GenericStore[T, PT]) existsInCache(ctx context.Context, id string) (bool, error) {
+	if !s.useCache {
+		return false, nil
+	}
+	s.cacheLock.RLock()
+	defer s.cacheLock.RUnlock()
+	val, found := s.cache[id]
+	found = found && s.isReadAllowed(ctx, val)
+	return found, nil
+}
+
+func (s *GenericStore[T, PT]) walkCache(ctx context.Context, walkFn func(obj PT) error) error {
+	if !s.useCache {
+		return nil
+	}
+	s.cacheLock.RLock()
+	defer s.cacheLock.RUnlock()
+	for _, v := range s.cache {
+		if !s.isReadAllowed(ctx, v) {
+			continue
+		}
+		err := walkFn(v)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *GenericStore[T, PT]) countFromCache(ctx context.Context) (int, error) {
+	if !s.useCache {
+		return 0, nil
+	}
+	count := 0
+	err := s.walkCache(ctx, func(_ PT) error {
+		count++
+		return nil
+	})
+	if err != nil {
+		return 0, err
+	}
+	return count, nil
+}
+
+func (s *GenericStore[T, PT]) isReadAllowed(ctx context.Context, obj PT) bool {
+	if s.hasPermissionsChecker() {
+		allowed, err := s.permissionChecker.ReadAllowed(ctx)
+		if err != nil {
+			return false
+		}
+		return allowed
+	}
+	scopeChecker := sac.GlobalAccessScopeChecker(ctx)
+	scopeChecker = scopeChecker.AccessMode(storage.Access_READ_ACCESS)
+	scopeChecker = scopeChecker.Resource(s.targetResource)
+	switch s.targetResource.GetScope() {
+	case permissions.NamespaceScope:
+		var interfaceObj interface{}
+		interfaceObj = obj
+		namespaceScopedObj := interfaceObj.(sac.NamespaceScopedObject)
+		scopeChecker = scopeChecker.ForNamespaceScopedObject(namespaceScopedObj)
+	case permissions.ClusterScope:
+		var interfaceObj interface{}
+		interfaceObj = obj
+		clusterScopedObj := interfaceObj.(sac.ClusterScopedObject)
+		scopeChecker = scopeChecker.ForClusterScopedObject(clusterScopedObj)
```

Comment:
Note: `storage.Cluster` is cluster-scoped but does not have a `GetClusterId` method.